### PR TITLE
Feature/epics acm

### PR DESCRIPTION
**Pull request by jluhrs**
First release of my Action-Command Model package for EPICS.

### CODE REVIEW

**Comment by swalker2m** [comment]
Since this is a rather large project (~4.6 KLOC), I think it would be helpful to have a bit of context.  Could you put together something in the wiki to illustrate the overall design of this code?  For example, I'm trying to document my [merge rework](https://github.com/gemini-hlsw/ocs/wiki/Science-Program-Synchronization-Notes) to use as a guide since it is also a large project that is difficult for someone not involved in it to follow.

**Comment by tpolecat** [comment]
Or perhaps as a `README.md` for the bundle itself? This way it won't get separated from the wiki?

**Comment by cquiroz** [comment]
Thank you for the documentation, love the ASCII art
There is mention that `CaCommandMonitorImpl` acts as a `Future`. Why not make it a future
I think scala `Future` or scalaz `Task` would be an excellent choice to implement Seqexec command sequencing

File: bundle/edu.gemini.epics.acm/src/test/java/edu/gemini/epics/acm/test/CaStatusAcceptorTest.java [hunk 1]
```diff
@@ -0,0 +1,179 @@
...
+	}
+
+	@Test
+	public void testGetStatusAcceptor() {
+		CaStatusAcceptor sa1 = caService.createStatusAcceptor(SA_NAME);
+		CaStatusAcceptor sa2 = caService.getStatusAcceptor(SA_NAME);
+
+		assertEquals("Retrieved the wrong CaStatusAcceptor.", sa1, sa2);
+        
+        caService.destroyStatusAcceptor(SA_NAME);
```
Comment (cquiroz) [inline bundle/edu.gemini.epics.acm/src/test/java/edu/gemini/epics/acm/test/CaStatusAcceptorTest.java]: Indentation seems off in some of these calls

File: bundle/edu.gemini.epics.acm/src/main/java/edu/gemini/epics/acm/CaCommandSenderImpl.java [hunk 2]
```diff
@@ -0,0 +1,295 @@
...
+                || stringParameters.containsKey(name)) {
+            return true;
+        } else {
+            return false;
+        }
+    }
+
+    @Override
+    public void mark() throws TimeoutException {
+        if (dirChannel != null) {
```
Comment (cquiroz) [inline bundle/edu.gemini.epics.acm/src/main/java/edu/gemini/epics/acm/CaCommandSenderImpl.java]: There are several instances of null checks that could be removed if the null were forbidden right at the constructor and final, unless of course null is a valid value

Comment (swalker2m) [inline bundle/edu.gemini.epics.acm/src/main/java/edu/gemini/epics/acm/CaCommandSenderImpl.java]: Also, we could simplify `alreadyExist` a bit:

````java
private boolean alreadyExist(String name) {
  return doubleParameters.containsKey(name) ||
                 floatParameters.containsKey(name) ||
                 integerParameters.containsKey(name) ||
                 stringParameters.containsKey(name);
}
````

**Comment by tpolecat** [comment]
I'm going to spend some time looking at this today, please hold off on merging.

**Comment by tpolecat** [comment]
Note that this bundle doesn't build unless you're on the Gemini network or VPN. So before merging we probably want to pull the bundle jarfiles out of `~/.ivy2` and put them in `ocs/lib/bundle` and put the `xmlunit` thing in the top-level `build.sbt` test dependencies.

**Comment by tpolecat** [comment]
I suggest running IDEA inspections and fixing the obvious issues. It found 575 items under the inspection settings I am using (mostly about things that can be final). You can put these in a file and import them in the inspections dialog.

```xml
<?xml version="1.0" encoding="UTF-8"?>
<inspections version="1.0" is_locked="false">
  <option name="myName" value="Project Default" />
  <option name="myLocal" value="false" />
  <inspection_tool class="CloneDeclaresCloneNotSupported" enabled="false" level="WARNING" enabled_by_default="false" />
  <inspection_tool class="FieldMayBeFinal" enabled="true" level="WARNING" enabled_by_default="true" />
  <inspection_tool class="JavaDoc" enabled="false" level="WARNING" enabled_by_default="false">
    <option name="TOP_LEVEL_CLASS_OPTIONS">
      <value>
        <option name="ACCESS_JAVADOC_REQUIRED_FOR" value="none" />
        <option name="REQUIRED_TAGS" value="" />
      </value>
    </option>
    <option name="INNER_CLASS_OPTIONS">
      <value>
        <option name="ACCESS_JAVADOC_REQUIRED_FOR" value="none" />
        <option name="REQUIRED_TAGS" value="" />
      </value>
    </option>
    <option name="METHOD_OPTIONS">
      <value>
        <option name="ACCESS_JAVADOC_REQUIRED_FOR" value="none" />
        <option name="REQUIRED_TAGS" value="@return@param@throws or @exception" />
      </value>
    </option>
    <option name="FIELD_OPTIONS">
      <value>
        <option name="ACCESS_JAVADOC_REQUIRED_FOR" value="none" />
        <option name="REQUIRED_TAGS" value="" />
      </value>
    </option>
    <option name="IGNORE_DEPRECATED" value="false" />
    <option name="IGNORE_JAVADOC_PERIOD" value="true" />
    <option name="IGNORE_DUPLICATED_THROWS" value="false" />
    <option name="IGNORE_POINT_TO_ITSELF" value="false" />
    <option name="myAdditionalJavadocTags" value="" />
  </inspection_tool>
  <inspection_tool class="LocalCanBeFinal" enabled="true" level="WARNING" enabled_by_default="true">
    <option name="REPORT_VARIABLES" value="true" />
    <option name="REPORT_PARAMETERS" value="true" />
  </inspection_tool>
  <inspection_tool class="LoggerInitializedWithForeignClass" enabled="false" level="WARNING" enabled_by_default="false">
    <option name="loggerClassName" value="org.apache.log4j.Logger,org.slf4j.LoggerFactory,org.apache.commons.logging.LogFactory,java.util.logging.Logger" />
    <option name="loggerFactoryMethodName" value="getLogger,getLogger,getLog,getLogger" />
  </inspection_tool>
  <inspection_tool class="NonJREEmulationClassesInClientCode" level="ERROR" enabled="false" />
  <inspection_tool class="SimplifiableIfStatement" enabled="false" level="WARNING" enabled_by_default="false" />
  <inspection_tool class="SpellCheckingInspection" enabled="false" level="TYPO" enabled_by_default="false" />
  <inspection_tool class="SuspiciousNameCombination" enabled="false" level="WARNING" enabled_by_default="false">
    <group names="x,width,left,right" />
    <group names="y,height,top,bottom" />
  </inspection_tool>
</inspections>

```

**Comment by fnussber** [comment]
Regarding @tpolecat s comment "pull the bundle jarfiles out of ~/.ivy2 and put them in ocs/lib/bundle". Wouldn't it be a good time to spend some effort on getting our infrastructure up-to-date so that we can start using our Gemini repository for sharing gemini artifacts (including with people outside the Gemini network)?

**Comment by tpolecat** [comment]
@fnussber this sounds good to me. I'm not sure what it would take to make this happen. Need to open a port and ensure the service is secured. I have never used it.

**Comment by swalker2m** [comment]
In the meantime, let's not add dependencies on our artifactory repository because they make it painful to build and work offsite and it goes against the spirit of having the code openly available.

**Comment by cquiroz** [comment]
In that case we should make a giapi release to have non-SNAPSHOT versions on lib

File: bundle/edu.gemini.epics.acm/src/test/java/edu/gemini/epics/acm/test/CaStatusAcceptorTest.java (hunk 1 above)
Comment (jluhrs) [inline bundle/edu.gemini.epics.acm/src/test/java/edu/gemini/epics/acm/test/CaStatusAcceptorTest.java]: Yes, my editor was using spaces, but the Autoformatter was set to use tabs.

File: bundle/edu.gemini.epics.acm/src/main/java/edu/gemini/epics/acm/CaCommandSenderImpl.java (hunk 2 above)
Comment (jluhrs) [inline bundle/edu.gemini.epics.acm/src/main/java/edu/gemini/epics/acm/CaCommandSenderImpl.java]: Duh!

**Comment by jluhrs** [comment]
I think I have fixed all the mayor issues in this PR. If no one disagree, I'll merge the PR next Tuesday (3/31)

**Comment by cquiroz** [comment]
Just out of curiosity, how are @tpolecat commits on the middle of the PR, they are related to acm though

**Comment by tpolecat** [comment]
@cquiroz I made a PR to @jluhrs's fork to straighten out his build setup.

**Comment by jluhrs** [comment]
Yes, but I just noticed that the history graph of the branch is in a bad state. A lots of commits appear twice, once in the branch and once in Rob's branch. The content and commit messages are the same, but the the commit ids are different. I may have messed thing when pulling from my working copy to get Rob's changes. Any idea how to fix it?

**Comment by tpolecat** [comment]
Guessing it was a5d97d3bde19e089d742dc96ddf0650a237c4c46 ... I don't really care as long as the end result is ok. Others may have a different opinion.

**Comment by jluhrs** [comment]
I rebased the branch, now history looks cleaner.

**Comment by tpolecat** [comment]
There we go.

**Pull request merged**